### PR TITLE
Fix Issue #36:  Avoid deleting uncommitted changes

### PR DESCRIPTION
**Pull request by BrenoJesusFernandes**
Basically solves https://github.com/pandas-dev/pandas-stubs/issues/36, when we run local tests now all changes are temporarily committed and restored after the tests. If there are not any new changes it is just gonna ignore.

### CODE REVIEW

**Comment by BrenoJesusFernandes** [comment]
> Would it make sense to skip mypy&pyright on the installed pandas-stubs when there are unstaged or uncommited changes (or even new files that are not tracked by git)?

I don't think so, and you @Dr-Irv ?

**Comment by twoertwein** [comment]
The reason for all of the git commands is that you want to make sure that when you run mypy&pyright on the installed pandas-stubs, that it doesn't use the local pandas-stubs, or? Instead of changing any files/git, how about changing into the `tests` folder,  overwrite the mypy_path/stubPath to use the current folder (which doesn't contain any stubs; in the worst case you still have to alter a file, but it is only pyproject.toml), mypy&pyright then cannot use the local stubs anymore, and then run `mypy .` and `pyright .` (which hopefully finds the installed stubs)?

**Comment by Dr-Irv** [comment]
Here's another suggestion. Instead of doing this with `git`, what we really care about is that the local copy of the `pandas-stubs` folder is not used when we test an installed distribution.  So how about we just temporarily rename `pandas-stubs` to `tempfile.TemporaryDirectory` and after the distribution is tested, you just rename it back.

**Comment by twoertwein** [comment]
> Here's another suggestion. Instead of doing this with `git`, what we really care about is that the local copy of the `pandas-stubs` folder is not used when we test an installed distribution. So how about we just temporarily rename `pandas-stubs` to `tempfile.TemporaryDirectory` and after the distribution is tested, you just rename it back.

I think that is by far the easiest method! Would need to be careful to rename the folder back to pandas-stubs (even in case of an error), otherwise the `tempfile.TemporaryDirectory` will get deleted. I would also be happy to just rename `pandas-stubs` to `_pandas-stubs` (and error if the folder already exists).

**Comment by BrenoJesusFernandes** [comment]
That's it, Done. Thx for the support guys!!

File: scripts/test/procedures.py [hunk 1]
```diff
@@ -31,8 +29,9 @@ def install_dist():
     subprocess.run(cmd, check=True)
 
 
-def remove_src():
-    shutil.rmtree(r"pandas-stubs")
+def rename_src():
+    if Path(r"pandas-stubs").exists():
```
Comment (Dr-Irv) [inline scripts/test/procedures.py:33]: can you abort here if for some reason `_pandas-stubs` already exists?

File: scripts/test/procedures.py [hunk 2]
```diff
@@ -44,14 +43,15 @@ def run_pyright_dist():
...
+
 def uninstall_dist():
     cmd = ["pip", "uninstall", "-y", "pandas-stubs"]
     subprocess.run(cmd, check=True)
 
 
 def restore_src():
-    cmd = ["git", "checkout", "HEAD", "pandas-stubs"]
-    subprocess.run(cmd, check=True)
+    if Path(r"_pandas-stubs").exists():
```
Comment (Dr-Irv) [inline scripts/test/procedures.py:53]: Can you abort here if `_pandas-stubs` somehow already exists?

File: scripts/test/__init__.py [hunk 3]
```diff
@@ -38,7 +39,7 @@ def test_dist(clean_cache: bool = False):
     steps.extend([
```
Comment (Dr-Irv) [inline scripts/test/__init__.py:39]: These steps are a subset of what is in `test_all`, so can you create a function for adding these steps that avoids the repeated code in `test_all()` and `test_dist()` ?

File: scripts/test/__init__.py [hunk 4]
```diff
@@ -62,11 +63,11 @@ def test_all(clean_cache: bool = False):
         Step(name="Run Pytest Against Source Code", run=procedures.run_pytest_src),
```
Comment (Dr-Irv) [inline scripts/test/__init__.py:63]: Similar to my remark about the `test_dist()` steps, the 3 steps here are the same as in `test_src()`, so can you set things up to only define each set of steps in one place and then use them accordingly?

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
See comments to make this a bit cleaner and more error-proof

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -52,8 +52,8 @@ jobs:
     - name: Install Distribution
       run: poetry run poe install_dist
 
-    - name: Remove Source Code
-      run: poetry run poe remove_src
+    - name: Rename Source Code
+      run: poetry run poe rename_src
```
Comment (twoertwein) [inline .github/workflows/test.yml:56]: Optional: would be nice to have it as a context manager

with rename_src():
    tests()

Comment (BrenoJesusFernandes) [reply .github/workflows/test.yml:56]: I did not get it, could you explain more details?

File: scripts/test/procedures.py (hunk 2 above)
Comment (BrenoJesusFernandes) [reply scripts/test/procedures.py:55]: Done

File: scripts/test/__init__.py (hunk 3 above)
Comment (BrenoJesusFernandes) [reply scripts/test/__init__.py]: Done

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
This looks good.  Can you merge with main and resolve conflicts?

**Comment by BrenoJesusFernandes** [comment]
> This looks good. Can you merge with main and resolve conflicts?

Sure

**Comment by BrenoJesusFernandes** [comment]
Please do not merge this PR now, for some reason pyupgrade is not working in my macos setup, I gonna test in other OS's to see what's going on.

**Review by Dr-Irv — APPROVED** [review]
I'm good with what is in here now.  Let me know if I can merge once the tests pass

**Comment by BrenoJesusFernandes** [comment]
> I'm good with what is in here now. Let me know if I can merge once the tests pass

Everything is ok now.

**Comment by Dr-Irv** [comment]
Thanks @BrenoJesusFernandes

**Pull request merged**